### PR TITLE
feat: acl cache discard

### PR DESCRIPTION
**Pull request by sh3d2**
## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## More information

Docs are not added yet, will be discussed in followup PRs for restapi / cli changes.

### CODE REVIEW

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:266]:
```suggestion
    _ = erlang:put(discard_k(), CurrentEvictionStamp),
    ok.
```

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:119]:
```suggestion
    ok = update_discarded_stamp().
```

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:259]:
```suggestion
is_discarded() ->
```

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:71]: We try to avoid nested `case` expressions to keep sane indentation levels.

If we change this function to accept a new args:
```suggestion
get_acl_cache(PubSub, Topic, _IsDiscarded = true)  ->
    _ = empty_ack_acche(),
    not_found;
```
then and keep the `false` clause untouched.

Which implies we need to delegate the /2 call to /3 like this:

```
get_acl_cache(PubSub, Topic) -> get_acl_cache(PubSub, Topic, is_discarded()).
```

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:262]: We should return `false | {true, NewStamp}` from this function.
And `NewStamp` should be passed to `update_discarded_stamp/1` instead of reading it again.

This is to avoid surprises when the stamp changes between these two calls.

**Review by zmstone — COMMENTED** [review]
Thank you for your PR @sh3d2

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:23]: I don't think we need to export this as it's only used locally.

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:51]: use a timestamp here ?

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:90]: delete empty line.

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:268]: order `true` clause before `false`.

**Review by zmstone — COMMENTED** [review]
Some nitpicking comments. :)

File: src/emqx_acl_cache.erl
Comment (sh3d2) [reply src/emqx_acl_cache.erl:51]: could you elaborate more?

Comment (zmstone) [reply src/emqx_acl_cache.erl:50]: sorry, brainfart. this is the key, which does not include the value (timestamp)

**Comment by zmstone** [comment]
maybe it's in your plan already: there should be an API exported e.g. `put_discarded_stamp` ?
which will be called from REST API handler and CLI handler.

**Comment by zmstone** [comment]
Please add new tests to cover discard logic in `test/emqx_acl_cache_SUITE.erl`

File: test/emqx_acl_cache_SUITE.erl
Comment (sh3d2) [inline test/emqx_acl_cache_SUITE.erl:59]: i have no idea if that kind of a test is sufficient. any feedback appreciated

**Review by sh3d2 — COMMENTED** [review]
> maybe it's in your plan already: there should be an API exported e.g. `put_discarded_stamp` ?
> which will be called from REST API handler and CLI handler.

should it be in `emqx_acl_cache` when this process is tied to connection? 
i guess it should be somewhere higher up the process tree, just i dunno where yet

File: test/emqx_acl_cache_SUITE.erl
Comment (zmstone) [reply test/emqx_acl_cache_SUITE.erl:59]: This should be enough IMO.
BTW this is not a `unit` test (as you stated in commit message).
In Erlang, unit is `eunit` (which are typically implemented in `_tests.erl` modules, or locally in the module under test).
Test cases in `_SUITE.erl` modules are `common` tests, or `ct`

Comment (sh3d2) [reply test/emqx_acl_cache_SUITE.erl:59]: ahh, good to know ;-)

File: src/emqx_acl_cache.erl
Comment (k32) [inline src/emqx_acl_cache.erl:258]: Disclaimer: I am not familiar with the code. But looking at the diff, perhaps a simpler implementation is possible, that doesn't require `is_discarded`. Instead, change implementation of `if_expired` to:

```erlang
case CachedAt =< EvictionTimestamp andalso (CachedAt + TTL) =< Now of
...
```
where `EvictionTimestamp` is obtained from the persistent term. It contains current time instead of a counter. There's a downside, though, that currently this module doesn't use monotonic time, so in theory it will be possible that node timewarps into the past, and starts dropping all caches.

**Comment by zmstone** [comment]
> > maybe it's in your plan already: there should be an API exported e.g. `put_discarded_stamp` ?
> > which will be called from REST API handler and CLI handler.
> 
> should it be in `emqx_acl_cache` when this process is tied to connection?
> i guess it should be somewhere higher up the process tree, just i dunno where yet

Yes, the `put_discarded_stamp` API (which calls `persistent_term` put) should be added to `emqx_acl_cache`.

File: src/emqx_acl_cache.erl
Comment (zmstone) [reply src/emqx_acl_cache.erl:258]: `EvictionTimestamp` serves as a unique marker, rather not used as an actual timestamp.

The idea is to set the stamp from any process,
then all the connection process can check if the stamp has been changed since they last time.
if yes, discard all cache.

Comment (zmstone) [reply src/emqx_acl_cache.erl:258]: I mean the timestamp compare maybe works OK. but it's confusing.

Comment (sh3d2) [reply src/emqx_acl_cache.erl:258]: Actually @k32 might be right. if we changed `EvictionTimestamp` role from unique marker to a timestamp, then (checking in `if_expired`) all cache entries added before `EvictionTimestamp` would become automagically evicted. 
Up to you @zmstone which way we go, im just learning here :-)

Comment (zmstone) [reply src/emqx_acl_cache.erl:258]: A second thought, since is_expired is down in the stack (except for dump_acl_cache which does not respect TTL anyway), we can avoid a PD put to remember the last stamp.

@sh3d2, please change it as @k32 suggested.

**Comment by zmstone** [comment]
> > > > maybe it's in your plan already: there should be an API exported e.g. `put_discarded_stamp` ?
> > > > which will be called from REST API handler and CLI handler.
> > > 
> > > 
> > > should it be in `emqx_acl_cache` when this process is tied to connection?
> > > i guess it should be somewhere higher up the process tree, just i dunno where yet
> > 
> > 
> > Yes, the `put_discarded_stamp` API (which calls `persistent_term` put) should be added to `emqx_acl_cache`.
> 
> could we just export `update_discarded_stamp/0`?

It's `update_discarded_stamp/1`, not `/0`
and in the new implementation as @k32 suggested, we do not need this update.
so maybe just rename it to `put_discarded_stamp/0` and export it.
i.e. when RPC call from another node, it should take its local timestamp, not from the caller node.

**Comment by k32** [comment]
Perhaps have two API functions: `start_cache_drain()` that puts local time to the PT, and `stop_cache_drain()` that deletes it?

**Comment by sh3d2** [comment]
> Perhaps have two API functions: `start_cache_drain()` that puts local time to the PT, and `stop_cache_drain()` that deletes the PT?

`start_cache_drain/0` sounds so much better than `put_discarded_stamp` ;-) 
i just cant see any use for `stop_cache_drain`. on the other hand, if youre able to start something, it would be good to have some ability to stop it (whatever it means).

File: src/emqx_acl_cache.erl
Comment (k32) [inline src/emqx_acl_cache.erl:32]: Small nitpick: the closing square bracket is typically indented at the same level as the previous comma

File: src/emqx_acl_cache.erl
Comment (k32) [inline src/emqx_acl_cache.erl:52]: Nit: Renaming the key `discard_stamp -> drain_stamp` will make it more greppable in the code

**Review by k32 — APPROVED** [review]

**Comment by k32** [comment]
LGTM, but someone else, who is perhaps more familiar with the code, should approve.

**Comment by sh3d2** [comment]
after this is merged im gonna move to `emqx-management` plugin in order to implement rest/cli commands

File: src/emqx_acl_cache.erl
Comment (k32) [inline src/emqx_acl_cache.erl:151]: Please add comments what the new functions do

File: src/emqx_acl_cache.erl
Comment (sh3d2) [inline src/emqx_acl_cache.erl:157]: i think this function does absolutely nothing as when `start_cache_drain()` is invoked all previous cache items disappear, items added after that stay no matter `stop_cache_drain()` is invoked or not.

i propose renaming `start_cache_drain` to `drain_cache` and remove `stop_cache_drain` completely

Comment (zmstone) [reply src/emqx_acl_cache.erl:157]: agree.

**Comment by zmstone** [comment]
@sh3d2 do you plan to add REST API and CLI handlers in another PR? If yes, this looks good to merge.

@HJianBo @terry-xiaoyu Maybe you guys should have a look. ?

**Comment by sh3d2** [comment]
> @sh3d2 do you plan to add REST API and CLI handlers in another PR? If yes, this looks good to merge.

yup, hopefully i will find some time for that tomorrow.

question: 
do we need anyting else in `emqx` repository? 
cli and rest is managed inside `emqx-management` plugin and my idea was to add the rest of the code there

**Comment by zmstone** [comment]
> > @sh3d2 do you plan to add REST API and CLI handlers in another PR? If yes, this looks good to merge.
> 
> yup, hopefully i will find some time for that tomorrow.
> 
> question:
> do we need anyting else in `emqx` repository?
> cli and rest is managed inside `emqx-management` plugin and my idea was to add the rest of the code there
> 
> ps. that was the most extensive and fun cr i had in a while, thanks!

hey. I’m glad you enjoyed it.
This is now a monorepo (or in rebar term, an umbrella project) at `master` branch.
Management plugin is right here in `apps` dir

**Comment by sh3d2** [comment]
> This is now a monorepo (or in rebar term, an umbrella project) at `master` branch.
> Management plugin is right here in `apps` dir

Ah, if thats the case theres no point creating another PR. im gonna carry on with rest/cli handlers

**Comment by zmstone** [comment]
rest/cli handlers do not seem quite tangled with the changes in this PR.

I actually prefer to have it merged now if you can rebase it the tidy up the commits a bit.

**Comment by sh3d2** [comment]
> I actually prefer to have it merged now if you can rebase it to tidy up the commits a bit.

commit history cleared. after we've merged this one, i will start a new pr with rest/cli handlers

**Review by zmstone — APPROVED** [review]

**Review by zmstone — DISMISSED** [review]

File: src/emqx_acl_cache.erl
Comment (zmstone) [inline src/emqx_acl_cache.erl:150]:
```suggestion
%% All acl cache entries added before `drain_cache()` invocation will become expired
```

Comment (zmstone) [reply src/emqx_acl_cache.erl:150]: Sorry, approved too soon.
one last bit.

Comment (sh3d2) [reply src/emqx_acl_cache.erl:150]: geez stupid me. fixing ;)

Comment (sh3d2) [reply src/emqx_acl_cache.erl]: done

**Review by zmstone — APPROVED** [review]

**Comment by zmstone** [comment]
Thank you for your contribution @sh3d2

**Pull request merged**
